src/Microsoft.VisualStudio.Threading/ThreadingTools.cs

```diff
@@ -143,6 +143,67 @@ public static Task WithCancellation(this Task task, CancellationToken cancellati
         return WithCancellationSlow(task, continueOnCapturedContext: false, cancellationToken: cancellationToken);
     }
 
+    /// <summary>
+    /// Wait a long running or later finishing task, but abort if this work is blocking the main thread.
+    /// </summary>
+    /// <param name="context">The JoinableTaskContext.</param>
+    /// <param name="slowTask">A slow task to wait.</param>
+    /// <param name="cancellationToken">An optional cancellation token.</param>
+    /// <returns>A task is completed either the slow task is completed, or the input cancellation token is triggered, or the context task blocks the main thread (inside JTF.Run).</returns>
+    /// <exception cref="OperationCanceledException">Throw when the cancellation token is triggered or the current task blocks the main thread.</exception>
+    public static Task WaitUnlessBlockingMainThreadAsync(this JoinableTaskContext context, Task slowTask, CancellationToken cancellationToken = default)
```

Comment:
Consider a slightly different signature to be similar to WithCancellation():
CancelIfMainThreadBlockedAsync(this Task slowTask, JoinableTaskContext context, CT)

Reply:
I like the order of parameters suggested by Richard, and updated the PR. I would like to hear more feedback on the name of the method itself, so I haven't updated it yet.